### PR TITLE
[FIX] service db: Restore database filestore in proper location.

### PR DESCRIPTION
**Pull request by Jerther**
**Description of the issue/feature this PR addresses:**
Database dumps (at least those received from upgrade.odoo.com) may contain a filestore folder. When restoring using the web database manager, this folder is not restored properly, and then Odoo complains about missing ir_attachment files. In my case, avatars were missing as well as many app switcher icons.

**Current behavior before PR:**
Filestore is restored in .../.local/.share/Odoo/filestore/db_name/filestore

**Desired behavior after PR is merged:**
Filestore is restores in .../.local/share/Odoo/filestore/db_name

**Additional details:**
The problem is that Odoo moves the filestore folder itself from the dump to the filestore folder.

The fix suggested here seemed overkill to me at first, but there's a situation where a simple operation with shutil would fail because at the time Odoo tries to move the filestore, it already contains at least a non-empty folder named "checklist" and shutil.move() fails to overwrite it. Plus, we want to keep its content. So I believe a move/merge hybrid is in order.

This will move each file from the dump to the filestore, creating new folders and files, overwriting only files, and then delete the source folder (which, at that point, contains only empty folders). This last step could be removed since the source folder is temporary and will be deleted later in the restore process anyway.

Maybe a copy tree solution would have been simpler, but it would be slower and could end up duplicating gigabytes.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by gustavovalverde** [comment]
Confirming this issue @JKE-be @Yenthe666 "

@Jerther, did you also create an issue with the enterprise code?

**Comment by Jerther** [comment]
You mean a github issue? If so, I can't, since only Odoo partners can access it.

**Comment by gustavovalverde** [comment]
@Jerther Ok..I'll report it as I'm on the same boat.

**Comment by Yenthe666** [comment]
To be approved by @mart-e or @odony.

**Comment by Jerther** [comment]
ALSO: the quick help at the bottom of upgrade.odoo.com (question # 8) says the original filestore should be manually copied before restoring the database. This PR works with that also.

**Comment by mart-e** [comment]
> Current behavior before PR:
> Filestore is restored in .../.local/.share/Odoo/filestore/db_name/filestore

This will happen only if you already have a `.local/.share/Odoo/filestore/db_name/` folder. Shouldn't we handle this usecase (e.g. block or replace) instead? In which scenario does it makes sense to merge the filestore ?
Indeed, seems overkill.

**Comment by gustavovalverde** [comment]
@mart-e, I'm not understanding your statement. 

For example, if my filestore in my configuration file is `/odoo/odoo-server/data`, when I create a database odoo appends /odoo/odoo-server/data`/filestore/db_name`

When I send this database to [upgrade.odoo.com](url), they return me a dump which appends /odoo/odoo-server/data`/filestore/db_name/filestore` and tries to use that one.

If the code won't be changed, then the dumps that are being generated by the upgrade service have to.

**Comment by Jerther** [comment]
I agree with @gustavovalverde . When restoring a database, Odoo creates the filestore folder first, then it moves the filestore that's inside the database backup. So yes, it already exists.

To me It makes sens to merge it because upgrade.odoo.com says to copy the original folestore BEFORE restoring the database.

Note: upgrade.odoo.com returns a backup that may contain a filestore because some ir_attachment were moved to filestore in the upgrade process.

**Comment by gustavovalverde** [comment]
Or maybe this information is wrong?
image

As this won't actually work based on the delivered database.

**Comment by Jerther** [comment]
@gustavovalverde Hmm, this part works fine for me. Here's what I do:

- Copy the original filestore as instructed
- Restore the database with web interface, giving it the exact same name as the previous step

With the original Odoo code, I end up with all my original attachments working, but many app switcher icons and avatars are missing.
If I do the exact same thing but with this PR, everything works fine.

**Comment by gustavovalverde** [comment]
@Jerther Even though I can manage to have my attachments, my log complains with 40 entries every 10 seconds because of missing avatars, icons, etc.

**Comment by Jerther** [comment]
Avatars and icons were moved to filestore by upgrade.odoo.com and should be in the upgraded database dump but Odoo restores them in the wrong place. Can you try your upgrade process with the code from this PR/branch?

**Comment by mart-e** [comment]
Let's not talk about the upgrade service here please (if there is an issue with it, it should be discussed with the upgrade team).
The typical scenario should be 

1. make the dump from the database manager
2. restore it using the same manager

The issue is that, if there was a previous database (including a filestore) with the same name on the machine, it will create a subfolder `filestore` _inside_ the existing `filestore/db_name` folder.

Are we talking about the same issue ?

**Comment by Jerther** [comment]
@mart-e  Not exactly. If the databse already exists, Odoo refuses to proceede.

I confirm that this issue also occurs with a dump created with the database manager (archive contains a `filestore` folder)

At the point where Odoo restores the filestore from the database dump, the filestore is already created as part of the database creation/restoration process. That is:

- Hit restore on the web interface
- Create an empty database (This creates a new _non-empty_ filestore)
- Unzip database backup archive
- Restore data from dump.sql with psql
- If any, move `filestore` inside the existing `filestore/db_name` folder. <-- This should move `filestore` content, not the folder itself.

**Comment by mart-e** [comment]
> If the databse already exists, Odoo refuses to proceede.

Obviously. I meant: there was a database (so a filestore was present) and the database is no longer present _but_ the previous filestore is still present.

But it seems you are talking about a different issue.

> Create an empty database **(This creates a new non-empty filestore)**

This is the part I am not getting. Restoring using the database manager should [only create the database](https://github.com/odoo/odoo/blob/45e94ec756e0da4766110bd3ce94ede4203e4486/odoo/service/db.py#L77), not create a filestore. I tested on the latest 10.0 and got no issue on that.

Do you have some specific configuration that may do that?

**Comment by Jerther** [comment]
... Interresting... Everything works fine restoring a vanilla odoo10e database dump with demo data... very interresting... Thanks for the lead. Let me dig into this a bit further.

**Comment by Jerther** [comment]
I have narrowed down the initial creation of the filestore. It's [here](https://github.com/odoo/odoo/blob/10.0/odoo/service/db.py#L258) which leads [here](https://github.com/odoo/odoo/blob/10.0/odoo/modules/registry.py#L66).

Although both the archive from upgrade.odoo.com and the archive made from the web interface have very similar structure, only the one coming from upgrade.odoo.com cause the premature creation of the filestore at that line. Any thoughts?

**Comment by mart-e** [comment]
Thank you for the info. I think I have figured out what the issue is.
It should _not_ happen that you get a new filestore during the new registry.

However, if you try to access the server _during_ the restore of the database, you will initiate a filestore (for the assets of the page).
This could happen if you have another browser tab opened that is polling or trying to reconnect to the server.

Could you check your logs if you have requests _before_ the final log entry?
`odoo.service.db: RESTORE DB: <dbname>`

**Comment by Jerther** [comment]
Yes there are other entries. I did many restores yesterday, and they end about all the same:

```
...
2017-06-08 12:47:15,380 25696 INFO mesa odoo.models.unlink: User #1 deleted ir.model.data records with IDs: [12793]
2017-06-08 12:47:15,382 25696 INFO mesa odoo.models.unlink: User #1 deleted ir.ui.view records with IDs: [1074]
2017-06-08 12:47:15,384 25696 INFO mesa odoo.modules.loading: Modules loaded.
2017-06-08 12:47:15,826 25696 INFO mesa odoo.service.db: RESTORE DB: mesa
```
```
...
2017-06-08 15:02:40,730 25696 DEBUG deleteme odoo.modules.loading: loading module metal_procurement (102/103)
2017-06-08 15:02:40,731 25696 DEBUG deleteme odoo.modules.loading: loading module metal_rfq (103/103)
2017-06-08 15:02:40,733 25696 INFO deleteme odoo.modules.loading: 103 modules loaded in 5.85s, 0 queries
2017-06-08 15:02:41,977 25696 INFO deleteme odoo.modules.loading: Modules loaded.
2017-06-08 15:02:42,457 25696 INFO deleteme odoo.service.db: RESTORE DB: deleteme
```
```
...
2017-06-08 15:14:11,168 25696 INFO deleteme2 odoo.addons.base.ir.ir_model: Deleting 1074@ir.ui.view (metal_crm.view_res_partner_filter)
2017-06-08 15:14:11,178 25696 INFO deleteme2 odoo.models.unlink: User #1 deleted ir.model.data records with IDs: [12793]
2017-06-08 15:14:11,180 25696 INFO deleteme2 odoo.models.unlink: User #1 deleted ir.ui.view records with IDs: [1074]
2017-06-08 15:14:11,181 25696 INFO deleteme2 odoo.modules.loading: Modules loaded.
2017-06-08 15:14:11,561 25696 INFO deleteme2 odoo.service.db: RESTORE DB: deleteme2
```
```
...
7-06-08 15:29:30,727 20728 INFO deleteme2 odoo.addons.base.ir.ir_model: Deleting 1074@ir.ui.view (metal_crm.view_res_partner_filter)
2017-06-08 15:29:30,744 20728 INFO deleteme2 odoo.models.unlink: User #1 deleted ir.model.data records with IDs: [12793]
2017-06-08 15:29:30,746 20728 INFO deleteme2 odoo.models.unlink: User #1 deleted ir.ui.view records with IDs: [1074]
2017-06-08 15:29:30,747 20728 INFO deleteme2 odoo.modules.loading: Modules loaded.
2017-06-08 15:29:31,209 20728 INFO deleteme2 odoo.service.db: RESTORE DB: deleteme2
```
```
...
2017-06-08 16:29:22,030 20728 INFO deleteme2 odoo.addons.base.ir.ir_model: Deleting 1074@ir.ui.view (metal_crm.view_res_partner_filter)
2017-06-08 16:29:22,041 20728 INFO deleteme2 odoo.models.unlink: User #1 deleted ir.model.data records with IDs: [12793]
2017-06-08 16:29:22,042 20728 INFO deleteme2 odoo.models.unlink: User #1 deleted ir.ui.view records with IDs: [1074]
2017-06-08 16:29:22,043 20728 INFO deleteme2 odoo.modules.loading: Modules loaded.
2017-06-08 16:29:22,395 20728 INFO deleteme2 odoo.service.db: RESTORE DB: deleteme2
```

Could it be an all around module upgrade? Like [at this line](https://github.com/odoo/odoo/blob/10.0/odoo/modules/registry.py#L82) maybe?

**Comment by Jerther** [comment]
Also, when I start a restore for deleteme2, here are the first few lines about it in the log:

```
2017-06-08 15:13:39,143 25696 INFO deleteme2 odoo.modules.loading: loading 1 modules...
2017-06-08 15:13:39,150 25696 DEBUG deleteme2 odoo.modules.loading: loading module base (1/1)
2017-06-08 15:13:39,415 25696 INFO deleteme2 odoo.modules.loading: 1 modules loaded in 0.26s, 0 queries
2017-06-08 15:13:39,806 25696 DEBUG deleteme2 odoo.modules.module: module change_tree_view: no manifest file found ('__manifest__.py', '__openerp__.py')
2017-06-08 15:13:39,806 25696 WARNING deleteme2 odoo.modules.graph: module change_tree_view: not installable, skipped
...
2017-06-08 15:13:40,694 25696 DEBUG deleteme2 odoo.modules.module: module purchase_order_line_quantity: no manifest file found ('__manifest__.py', '__openerp__.py')
2017-06-08 15:13:40,694 25696 WARNING deleteme2 odoo.modules.graph: module purchase_order_line_quantity: not installable, skipped
2017-06-08 15:13:41,904 25696 ERROR deleteme2 odoo.modules.graph: module metal_account: Unmet dependencies: account_payment_mode, account_payment_partner
2017-06-08 15:13:41,904 25696 DEBUG deleteme2 odoo.modules.loading: Updating graph with 127 more modules
2017-06-08 15:13:41,917 25696 INFO deleteme2 odoo.modules.loading: loading 103 modules...
2017-06-08 15:13:41,918 25696 DEBUG deleteme2 odoo.modules.loading: loading module auth_crypt (2/103)
2017-06-08 15:13:41,920 25696 DEBUG deleteme2 odoo.modules.loading: loading module contacts (3/103)
2017-06-08 15:13:41,920 25696 DEBUG deleteme2 odoo.modules.loading: loading module decimal_precision (4/103)
2017-06-08 15:13:41,935 25696 DEBUG deleteme2 odoo.modules.loading: loading module resource (5/103)
...
2017-06-08 15:14:02,443 25696 DEBUG deleteme2 odoo.modules.loading: loading module metal_partner (95/103)
2017-06-08 15:14:02,896 25696 INFO deleteme2 odoo.modules.registry: module metal_partner: creating or updating database tables
2017-06-08 15:14:03,025 25696 INFO deleteme2 odoo.modules.loading: loading metal_partner/views/partner.xml
2017-06-08 15:14:03,320 25696 INFO deleteme2 odoo.addons.base.ir.ir_translation: module metal_partner: loading translation file (fr) for language fr_FR
2017-06-08 15:14:03,321 25696 INFO deleteme2 odoo.tools.translate: loading /opt/odoo/addons_local/enabled/metal_partner/i18n/fr.po
```

**Comment by Jerther** [comment]
If we're looking for other process that might access the database, could it be Odoo's cron engine? Since the restore can take up to 20 minutes, there are a few of these during the restore:

`2017-06-08 15:14:42,683 25696 DEBUG ? odoo.service.server: cron0 polling for jobs`

**Comment by mart-e** [comment]
Ok thanks for the information. I have made a patch to avoid this at #17531

**Comment by Jerther** [comment]
I don't know @mart-e . My logs clearly show entries relates to the database being restored, but none look like a web request such as:
`2017-06-05 15:32:09,123 8558 INFO ? werkzeug: 192.168.5.181 - - [05/Jun/2017 15:32:09] "GET /web_enterprise/static/src/img/mobile-icons/android-192x192.png HTTP/1.1" 200 -`

When restoring a database called deleteme2, the first two lines about it in the log are litterally:

```
2017-06-08 15:13:39,143 25696 INFO deleteme2 odoo.modules.loading: loading 1 modules...
2017-06-08 15:13:39,150 25696 DEBUG deleteme2 odoo.modules.loading: loading module base (1/1)
```

If it were a web call, shouldn't the log show it before those two lines?

**Comment by Jerther** [comment]
GOT IT.

### TL;DR:
The filestore can be created DURING the restore process and BEFORE it is restored from the dump. It happens when the database being restored contains installable modules in the "to upgrade" state, having files to store like menu icons.

I believe this PR still best fix this scenario.

### Explanation:
Here is [registry.new()](https://github.com/odoo/odoo/blob/10.0/odoo/modules/registry.py#L81-L82) method with some log added, which is called from [db.restore_db()](https://github.com/odoo/odoo/blob/10.0/odoo/service/db.py#L258).

```
# This should be a method on Registry
_logger.info('0 registry load_modules for db: %s, force_demo: %s, status: %s, update_module: %s' % (db_name, force_demo, status, update_module))
_logger.info('1 /var/lib/odoo/.local/share/Odoo/filestore/deleteme4 exists: %s' % os.path.exists('/var/lib/odoo/.local/share/Odoo/filestore/deleteme4'))
odoo.modules.load_modules(registry._db, force_demo, status, update_module)
_logger.info('2 /var/lib/odoo/.local/share/Odoo/filestore/deleteme4 exists: %s' % os.path.exists('/var/lib/odoo/.local/share/Odoo/filestore/deleteme4'))
```
[Here's a log file](https://github.com/odoo/odoo/files/1064444/odoo.txt) that starts at the moment where I hit restore in the web interface until the db is fully restored. Some highlights:
```
2017-06-09 14:48:24,255 24597 INFO None odoo.modules.registry: 0 registry load_modules for db: deleteme4, force_demo: False, status: None, update_module: False
2017-06-09 14:48:24,256 24597 INFO None odoo.modules.registry: 1 /var/lib/odoo/.local/share/Odoo/filestore/deleteme4 exists: False
2017-06-09 14:48:24,646 24597 INFO deleteme4 odoo.modules.loading: loading 1 modules...
...
2017-06-09 14:49:05,054 24597 INFO deleteme4 odoo.modules.loading: Modules loaded.
2017-06-09 14:49:05,065 24597 INFO deleteme4 odoo.modules.registry: 2 /var/lib/odoo/.local/share/Odoo/filestore/deleteme4 exists: True
2017-06-09 14:49:05,384 24597 INFO deleteme4 odoo.service.db: RESTORE DB: deleteme4
```
So the filestore **IS** created by some module **DURING** the restore process. In my case, the filestore created that way contains a single file so I've looked it up, and it's for a ir.ui.menu icon for a custom module I've made:

`<menuitem id="topo_main_menu" name="Topo" web_icon="topo,static/description/icon.png" groups="topo.users" />`

This corelates with modules triggering `creating or updating database tables` in the log, but only _Topo_ has an icon to store:

Digging down further (really want to get to the bottom of this), I've unzipped the database archive, and manually restored dump.sql, looked at the modules status and ALL `creating or updating database tables` are triggered by modules in the 'to upgrade' state and with installable = True.

Also, the added log lines are found only once during the restore process, ruling out the "external request to /web" possibility.

**Comment by mart-e** [comment]
> when the database being restored contains installable modules in the "to upgrade" state

Ok, I discussed with the upgrade team, this is part of the upgrade procedure for custom modules. This is indeed another way to trigger the creation of the filestore.

I still think it is not a good idea to merge the filestores and should instead postpone the upgrade of the modules at the end of the restoration but I let @odony give its opinion on this or #17531 solution.

**Comment by gustavovalverde** [comment]
So what should be the correct approach to restore the database (with filestore) without this or @mart-e 's PR (https://github.com/odoo/odoo/pull/17531), as I have to do this procedure today.

**Comment by Jerther** [comment]
Hmm, to me, "correct" means no manual intervention and I think it's not possible without changing some code. Anyway, here's a workaround:

- First, restore your database.
- Then, manually fix the filestore by moving the filestore/dbname/filestore content into filestore/dbname/, if any.
- Finally, copy your previous filestore as described in upgrade.odoo.com question number 8.

This is what I did before I made this PR.

**Comment by pedrobaeza** [comment]
Come to this through codetriage.com. What is the final verdict on this?

**Comment by mart-e** [comment]
I don't think there is a good solution without actually modifying the interface (and ask the user what to do, or block the restoration).
Let's not change it for now.
